<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Откройте-и-изучите-файл" data-toc-modified-id="Откройте-и-изучите-файл-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Откройте и изучите файл</a></span><ul class="toc-item"><li><span><a href="#Подключение-библиотек" data-toc-modified-id="Подключение-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подключение библиотек</a></span></li><li><span><a href="#Загрузка-файла" data-toc-modified-id="Загрузка-файла-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Загрузка файла</a></span></li><li><span><a href="#Функция-для-первичного-анализа-базы" data-toc-modified-id="Функция-для-первичного-анализа-базы-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Функция для первичного анализа базы</a></span></li><li><span><a href="#Анализ-базы" data-toc-modified-id="Анализ-базы-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Анализ базы</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Разбейте-данные-на-выборки" data-toc-modified-id="Разбейте-данные-на-выборки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбейте данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Формирование-признаков-features-и-целевого-признака-target" data-toc-modified-id="Формирование-признаков-features-и-целевого-признака-target-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Формирование признаков <code>features</code> и целевого признака <code>target</code></a></span></li><li><span><a href="#Разбивка-данных-на-обучающую,-валидационную-и-тестовую-выборки" data-toc-modified-id="Разбивка-данных-на-обучающую,-валидационную-и-тестовую-выборки-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Разбивка данных на обучающую, валидационную и тестовую выборки</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Модель-решающее-дерево" data-toc-modified-id="Модель-решающее-дерево-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Модель решающее дерево</a></span><ul class="toc-item"><li><span><a href="#Функция-подбора-гиперпараметров-для-Дерева-решений" data-toc-modified-id="Функция-подбора-гиперпараметров-для-Дерева-решений-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Функция подбора гиперпараметров для Дерева решений</a></span></li><li><span><a href="#Сохранение-модели-с-наилучшими-гиперпараметрами" data-toc-modified-id="Сохранение-модели-с-наилучшими-гиперпараметрами-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Сохранение модели с наилучшими гиперпараметрами</a></span></li><li><span><a href="#Перекрёстная-проверка" data-toc-modified-id="Перекрёстная-проверка-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Перекрёстная проверка</a></span></li></ul></li><li><span><a href="#Модель-случайный-лес" data-toc-modified-id="Модель-случайный-лес-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Модель случайный лес</a></span><ul class="toc-item"><li><span><a href="#Функция-подбора-гиперпараметров-для-Случайного-леса" data-toc-modified-id="Функция-подбора-гиперпараметров-для-Случайного-леса-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Функция подбора гиперпараметров для Случайного леса</a></span></li><li><span><a href="#Сохранение-модели-с-наилучшими-гиперпараметрами" data-toc-modified-id="Сохранение-модели-с-наилучшими-гиперпараметрами-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Сохранение модели с наилучшими гиперпараметрами</a></span></li><li><span><a href="#Перекрёстная-проверка" data-toc-modified-id="Перекрёстная-проверка-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Перекрёстная проверка</a></span></li></ul></li><li><span><a href="#Модель-логистическая-регрессия" data-toc-modified-id="Модель-логистическая-регрессия-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Модель логистическая регрессия</a></span><ul class="toc-item"><li><span><a href="#Функция-подбора-гиперпараметров-для-Логистической-регрессии" data-toc-modified-id="Функция-подбора-гиперпараметров-для-Логистической-регрессии-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Функция подбора гиперпараметров для Логистической регрессии</a></span></li><li><span><a href="#Сохранение-модели-с-наилучшими-гиперпараметрами" data-toc-modified-id="Сохранение-модели-с-наилучшими-гиперпараметрами-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Сохранение модели с наилучшими гиперпараметрами</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверьте-модель-на-тестовой-выборке" data-toc-modified-id="Проверьте-модель-на-тестовой-выборке-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверьте модель на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Проверка-модели-решающее-дерево" data-toc-modified-id="Проверка-модели-решающее-дерево-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Проверка модели решающее дерево</a></span></li><li><span><a href="#Проверка-модели-случайный-лес" data-toc-modified-id="Проверка-модели-случайный-лес-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Проверка модели случайный лес</a></span></li><li><span><a href="#Проверка-модели-логистическая-регрессия" data-toc-modified-id="Проверка-модели-логистическая-регрессия-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Проверка модели логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#(бонус)-Проверьте-модели-на-адекватность" data-toc-modified-id="(бонус)-Проверьте-модели-на-адекватность-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>(бонус) Проверьте модели на адекватность</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Рекомендация тарифов

Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Пороговое значение **accuracy** - не ниже 0,75.

# Описание данных
Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц.

* сalls — количество звонков,
* minutes — суммарная длительность звонков в минутах,
* messages — количество sms-сообщений,
* mb_used — израсходованный интернет-трафик в Мб,
* is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

## Загрузка и анализ данных

### Подключение библиотек

In [1]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score

### Загрузка файла

In [2]:
try:
    df = pd.read_csv('users_behavior.csv')
except:
    df = pd.read_csv('/datasets/users_behavior.csv')

### Функция для первичного анализа базы

In [3]:
def data_info(data):
    data.info()
    print('')
    print('*'*50)
    display(data.head())
    print('')
    print('*'*50)
    display(data.describe())

### Анализ базы

In [4]:
data_info(df)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB

**************************************************


calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0


**************************************************


calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

In [5]:
df['is_ultra'].value_counts()

0    2229
1     985
Name: is_ultra, dtype: int64

In [6]:
print('Процентное соотношение тарифов Смарт {:.0%}, Ультра {:.0%}'.format(len(df[df['is_ultra'] == 0])/len(df), 
                                                                          len(df[df['is_ultra'] == 1])/len(df)))

Процентное соотношение тарифов Смарт 69%, Ультра 31%


### Вывод

В нашем распоряжении предобработанные данные о 3214 пользователей, распределение пользователей по тарифам:
* "Смарт" - 2229 человек, **69%** от общего числа пользователей
* "Ультра" - 985 человек, **31%** от общего числа пользователей 

## Подготовка данных

### Формирование признаков `features` и целевого признака `target`

In [7]:
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

### Разбивка данных на обучающую, валидационную и тестовую выборки

In [8]:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.2, random_state=12345)
features_train, features_valid, target_train, target_valid = train_test_split(
    features_train, target_train, test_size=0.25, random_state=12345)
print('Размер обучающей выборки:', features_train.shape, 
      ', т.е.','{:.1%}'.format(len(features_train)/len(df)), 
      'от общего размера', end='\n')
print('Размер валидационной выборки:', features_valid.shape, 
      ', т.е.','{:.1%}'.format(len(features_valid)/len(df)), 
      'от общего размера', end='\n')
print('Размер тестовой выборки:', features_test.shape, 
      ', т.е.','{:.1%}'.format(len(features_test)/len(df)), 
      'от общего размера', end='\n')

Размер обучающей выборки: (1928, 4) , т.е. 60.0% от общего размера
Размер валидационной выборки: (643, 4) , т.е. 20.0% от общего размера
Размер тестовой выборки: (643, 4) , т.е. 20.0% от общего размера


### Вывод

Дважды воспользовавшись функцией `train_test_split` мы получили:
* **Обучающую** выборку, равную 60% от общего объема данных
* **Валидационную** выборку, равную 20% от общего объема данных
* **Тестовую** выборку, равную валидационной выборке и соответственно равной 20% от общего объема данных

## Обучение моделей

### Модель решающее дерево

#### Функция подбора гиперпараметров для Дерева решений

In [9]:
def tree(fear_train, target_train, fear_valid, target_valid, random_state, max_depth):
    best_depth = 0
    best_result = 0
    for depth in range(1, max_depth+1):
        model = DecisionTreeClassifier(max_depth = depth, random_state = random_state)
        model.fit(fear_train, target_train)
        prediction_valid = model.predict(fear_valid)
        result = accuracy_score(target_valid, prediction_valid)
        print('max_depth =', depth, ': ', end='')
        print('{:.3f}'.format(result))
        if result > best_result:
            best_depth = depth
            best_result = result
            model_tree = model
    print('-'*70)
    print('Наилучший результат Accuracy {:.3f} при максимальной глубине дерева {}'.format(best_result, best_depth))
    return model_tree

#### Сохранение модели с наилучшими гиперпараметрами

In [10]:
model_tree = tree(features_train, target_train, features_valid, target_valid, 12345, 6)

max_depth = 1 : 0.739
max_depth = 2 : 0.757
max_depth = 3 : 0.765
max_depth = 4 : 0.764
max_depth = 5 : 0.759
max_depth = 6 : 0.757
----------------------------------------------------------------------
Наилучший результат Accuracy 0.765 при максимальной глубине дерева 3


#### Перекрёстная проверка

In [11]:
print('Средняя оценка Accuracy модели перекрестной валидацией {:.2%}'
      .format(cross_val_score(model_tree, features, target, cv=5).mean()))

Средняя оценка Accuracy модели перекрестной валидацией 79.15%


### Модель случайный лес

#### Функция подбора гиперпараметров для Случайного леса

In [12]:
def forest(fear_train, target_train, fear_valid, target_valid, random_state, max_depth, n_est):
    best_depth = 0
    best_result = 0
    best_est = 0
    for est in range(1, n_est+1):
        print('n_est =', est)
        print('-'*30)
        for depth in range(1, max_depth+1):
            model = RandomForestClassifier(random_state = random_state, max_depth = depth, n_estimators = est)
            model.fit(fear_train, target_train)
            prediction_valid = model.predict(fear_valid)
            result = accuracy_score(target_valid, prediction_valid)
            print('max_depth =', depth, ': ', end='')
            print('{:.3f}'.format(result))
            if result > best_result:
                best_depth = depth
                best_result = result
                best_est = est
                model_forest = model
        print('')
        print('-'*30)
    print('Наилучший результат Accuracy {:.3f} при максимальной глубине дерева {} и максимальном количестве деревьев {}'
          .format(best_result, best_depth, best_est))
    return model_forest

#### Сохранение модели с наилучшими гиперпараметрами

In [13]:
model_forest = forest(features_train, target_train, features_valid, target_valid, 12345, 6, 15)

n_est = 1
------------------------------
max_depth = 1 : 0.726
max_depth = 2 : 0.757
max_depth = 3 : 0.757
max_depth = 4 : 0.770
max_depth = 5 : 0.771
max_depth = 6 : 0.767

------------------------------
n_est = 2
------------------------------
max_depth = 1 : 0.726
max_depth = 2 : 0.757
max_depth = 3 : 0.757
max_depth = 4 : 0.767
max_depth = 5 : 0.767
max_depth = 6 : 0.774

------------------------------
n_est = 3
------------------------------
max_depth = 1 : 0.726
max_depth = 2 : 0.757
max_depth = 3 : 0.760
max_depth = 4 : 0.764
max_depth = 5 : 0.764
max_depth = 6 : 0.779

------------------------------
n_est = 4
------------------------------
max_depth = 1 : 0.726
max_depth = 2 : 0.762
max_depth = 3 : 0.760
max_depth = 4 : 0.765
max_depth = 5 : 0.762
max_depth = 6 : 0.778

------------------------------
n_est = 5
------------------------------
max_depth = 1 : 0.726
max_depth = 2 : 0.736
max_depth = 3 : 0.760
max_depth = 4 : 0.767
max_depth = 5 : 0.770
max_depth = 6 : 0.785

------

#### Перекрёстная проверка

In [14]:
print('Средняя оценка Accuracy модели перекрестной валидацией {:.2%}'
      .format(cross_val_score(model_forest, features, target, cv=5).mean()))

Средняя оценка Accuracy модели перекрестной валидацией 80.34%


### Модель логистическая регрессия

#### Функция подбора гиперпараметров для Логистической регрессии

In [15]:
def regression(fear_train, target_train, fear_valid, target_valid, random_state, max_iter):
    best_result = 0
    best_iter = 0
    for iters in range(100, max_iter+1, 100):
        model = LogisticRegression(random_state = random_state, solver='lbfgs', max_iter=iters, multi_class='multinomial')
        model.fit(fear_train, target_train)
        prediction_valid = model.predict(fear_valid)
        result = model.score(fear_valid, target_valid)
        print('max_iter =', iters, ': ', end='')
        print('{:.3f}'.format(result))
        if result > best_result:
            best_iter = iters
            best_result = result
            model_reg = model
    print('-'*70)
    print('Наилучший результат Accuracy {:.3f} при максимальном количестве итераций {}'.format(best_result, best_iter))
    return model_reg

#### Сохранение модели с наилучшими гиперпараметрами

In [16]:
model_regression = regression(features_train, target_train, features_valid, target_valid, 12345, 1000)

max_iter = 100 : 0.726
max_iter = 200 : 0.726
max_iter = 300 : 0.726
max_iter = 400 : 0.726
max_iter = 500 : 0.726
max_iter = 600 : 0.726
max_iter = 700 : 0.726
max_iter = 800 : 0.726
max_iter = 900 : 0.726
max_iter = 1000 : 0.726
----------------------------------------------------------------------
Наилучший результат Accuracy 0.726 при максимальном количестве итераций 100


In [17]:
print('Средняя оценка Accuracy модели перекрестной валидацией {:.2%}'
      .format(cross_val_score(model_regression, features, target, cv=5).mean()))

Средняя оценка Accuracy модели перекрестной валидацией 72.71%


### Вывод

Рассмотрев три модели классификации, получены следующие результаты:
1. Модель Дерево решений
    * при разделении на обучающую и валидационную выборки имеет показатель `Accuracy` - **76,5%** при максимальной глубине дерева - 3
    * при перекрестной проверке по всему объему данных имеет `Accuracy` - **79,15%**
2. Модель Случайный лес 
    * при разделении на обучающую и валидационную выборки имеет показатель `Accuracy` - **79%** при максимальном количестве деревьев 6 и максимальной глубине дерева 6
    * при перекрестной проверке по всему объему данных имеет `Accuracy` - **80,34%**
3. Модель Логистическая регрессия 
    * при разделении на обучающую и валидационную выборки имеет показатель `Accuracy` - **72,6%** при максимальном количестве итераций - 100
    * при перекрестной проверке по всему объему данных имеет `Accuracy` - **72,71%**

## Проверка моделей на тестовой выборке

### Проверка модели решающее дерево

In [18]:
predict_tree = model_tree.predict(features_test)
result_tree = accuracy_score(target_test, predict_tree)
print('Accuracy модели дерево решений на тестовой выборке: {:.2%}'.format(result_tree))

Accuracy модели дерево решений на тестовой выборке: 78.69%


### Проверка модели случайный лес

In [19]:
predict_forest = model_forest.predict(features_test)
result_forest = accuracy_score(target_test, predict_forest)
print('Accuracy модели случайный лес на тестовой выборке: {:.2%}'.format(result_forest))

Accuracy модели случайный лес на тестовой выборке: 79.63%


### Проверка модели логистическая регрессия

In [20]:
predict_regression = model_regression.predict(features_test)
result_reg = model_regression.score(features_test, target_test)
print('Accuracy модели логистическая регрессия на тестовой выборке: {:.2%}'.format(result_reg))

Accuracy модели логистическая регрессия на тестовой выборке: 75.89%


### Вывод

При проверке моделей на тестовых данных, получены следующие результаты:
* Дерево решений - доля правильных ответов - **78,69%**
* Случайный лес - доля правильных ответов - **79,63%**
* Логистическая регрессия - доля правильных ответов - **75,89%**

## Проверка модели на адекватность

Так как мы имеем в данных распределение пользователей по тарифам *Смарт* и *Ультра* **69%** и **31%** соответственно, вероятность верно спрогнозировать тариф *Смарт* равна 69%, а для прогноза более редкого тарифа, в нашем случае это тариф "Ультра", метрика `Accuracy`, т.е. *доля правильных ответов*, должна быть больше 70%. 

Все рассмотренные модели классификации имеют показатель `Accuracy` больше **70%**, в связи с чем, можно судить об адекватности каждой рассмотренной модели классификации.

## Выводы

По результатам проведенного исследования, в ходе которого модели классификации оценивались только по признаку `Accuracy`, т.е. доли правильных ответов, наилучшим образом на тестовой выборке из общего объема данных себя показала модель **Случайного леса** имеющая показатель `Accuracy` равный **79,63%**.

Остальные рассмотренные модели так же справились с порогом в требуемые 75%:
* Модель **Дерево решений** достигла 78,69% на тестовой выборке
* Модель **Логистическая регрессия** достигла 75,89% на тестовой выборке

Для более точной оценки каждой модели необходимо так же рассмотреть такие критерии как:
* **Precision** - *точность*
* **Recall** - *полнота*
* **F1** - *баланс между точностью и полнотой*
* **ROC_AUC** - *площадь под кривой ошибок*